apps/rand.c

```diff
@@ -95,8 +97,85 @@ int rand_main(int argc, char **argv)
     argc = opt_num_rest();
     argv = opt_rest();
     if (argc == 1) {
-        if (!opt_long(argv[0], &num) || num <= 0)
+        int factoridx = 0;
+        int shift = 0;
+
+        /*
+         * special case for requesting the max allowed
+         * number of random bytes to be generated
+         */
+        if (!strcmp(argv[0], "max")) {
+            /*
+             * 2^61 bytes is the limit of random output
+             * per drbg instantiation
+             */
+            scaled_num = SIZE_MAX >> 3;
+        } else {
+            /*
+             * iterate over the value and check to see if there are
+             * any non-numerical chars
+             * A non digit suffix indicates we need to shift the
+             * number of requested bytes by a factor of:
+             * k = 1024^1 (1 << (10 * 1))
+             * m = 1024^2 (1 << (10 * 2))
+             * g = 1024^3 (1 << (10 * 3))
+             * t = 1024^4 (1 << (10 * 4))
+             * which can be achieved by bit-shifting the number
+             */
+            while (argv[0][factoridx]) {
+                if (!isdigit((int)(argv[0][factoridx]))) {
+                    switch(argv[0][factoridx]) {
+                    case 'K':
+                        shift = 10;
+                        break;
+                    case 'M':
+                        shift = 20;
+                        break;
+                    case 'G':
+                        shift = 30;
+                        break;
+                    case 'T':
```

Comment:
for platforms where sizeof(size_t) == 4, this case isn't required and will produce undefined behaviour according to _6.3.7 Bitwise shift operators_ in the C89 standard:

> If the value of the right operand is negative or is greater than or
equal to the width in bits of the promoted left operand. the behavior is undefined.

I.e. we should detect this case and error.

Something along the lines of
```c
if (shift / 8 >= sizeof(size_t))
    …
```
before line 165 where the shift is first done.

Reply:
I just noticed we don't have shift operations in safe_math.h. Could be nice to have in future.

Reply:
Good suggestion.  Both directional shifts could be added.  I'm not sure there is an error condition though.  Shifting too far would produce 0 (for unsigned) and 0 or -1 for signed.

Reply:
If we use uint64_t above, I think we avoid this problem entirely and lets us keep T, but I agree adding a safe left shift operation is a good idea.

Reply:
converted to a uint64